### PR TITLE
feat: Initial SQLite catalog schema

### LINKED ISSUE #6169: Feature request: Implement a Sqlite backed catalog
__Use case:__

1. I create a reproducer using non trivial amounts of data and/or steps on a local IOx instance (either for performance or for a crash issue)
2. I want to be able to change the code, recompile and test out fixes / changes locally (which may potentially change the catalog) without having to recreate the reproducer again

I also want to share the database in the broken state with another engineer so they can help debug whatever problem I am having.

# Context:

While working on https://github.com/influxdata/idpe/issues/16485 -- see the complicated reproducer setup in https://github.com/influxdata/idpe/issues/16485#issuecomment-1318746719

There are similar, hard to follow repro instructions on https://github.com/influxdata/idpe/issues/16298#issue-1427259604

__Proposal:__

Instead I would like to be able to run IOx against a sqlite file backed database which can be trivally moved from computer to computer

__Current behaviour:__
Technically I can save the database state today with local postgres instance and/or a docker image. However, doing so is cumbersome.

```shell
INFLUXDB_IOX_CATALOG_DSN=postgres://postgres@localhost:5432/postgres influxdb_iox
```

__Desired behaviour:__
We already use https://docs.rs/sqlx/latest/sqlx/ in iox

Currently we use the postgres driver: https://docs.rs/sqlx/latest/sqlx/postgres/index.html

sqlx  has a sqlite driver: https://docs.rs/sqlx/latest/sqlx/sqlite/index.html 

So I would like to be able to start iox like:

```shell
INFLUXDB_IOX_CATALOG_DSN=sqlite://data.db influxdb_iox run ingester
```

And have it use the sqlite database in `data.db`

https://docs.rs/sqlx/latest/sqlx/sqlite/struct.SqliteConnectOptions.html

__Alternatives considered:__

There are other potential ways of doing this (for example a of save/restore tool that extracts relevant entities from the catalog and object store into local files).

I don't see the two as mutually exclusive -- eventually it would be awesome to use sqlite as the format for the relevant catalog entities in a dump, for example

### PR DESCRIPTION
**Pull request by stuartcarnie**
Closes #6169

- [x] Create compatible SQLite schema
- [x] Implement Catalog
- [x] Add discriminator to allow SQLite DSN
- [ ] ~Integrate into end-to-end tests~
    - ~_Help needed_ here to understand all the touch points for introducing a new catalog type~
    - Will do in a follow-on PR

### Summary

Most of the SQL was reusable, with some minor modifications.

#### `INSERT .. ON CONFLICT` changes

SQLite cannot refer to a named index, so the columns must be listed explicitly:

```sql
INSERT .. ON CONFLICT (column1, column2, ...)
```

#### Binding to `Vec<_>`

SQLite does not support binding to parameters that are `Vec<_>`. This is used in PostgreSQL statements via the `UNNEST` and `ANY` statements. The SQLite catalog implementation handles this by using `sqlx`'s JSON type, which encodes the parameter to a JSON string. In turn the SQLite statement uses the `json_each` table-value function to decode the JSON.

https://github.com/influxdata/influxdb_iox/blob/90af3e238b5c8c36d1b80fd974b584bbb95928c4/iox_catalog/src/sqlite.rs#L1315-L1317

#### Private plain old datatype (POD) structures

A couple of types had to be specialised in the `sqlite` module, to customise the serialisation, as SQLite does not have support for array types. For example, the `PartitionPod` type specifies a `Json` field for the `sort_key`:

https://github.com/influxdata/influxdb_iox/blob/90af3e238b5c8c36d1b80fd974b584bbb95928c4/iox_catalog/src/sqlite.rs#L1003-L1011

The `sort_key` vector is serialised as a `TEXT` field, encoded as a JSON array of integers.

#### Triggers

SQLite doesn't have an `IF` statement, however, the `WHEN` clause provides the equivalent behaviour for all the PostgreSQL triggers:

https://github.com/influxdata/influxdb_iox/blob/90af3e238b5c8c36d1b80fd974b584bbb95928c4/iox_catalog/sqlite/migrations/20230204082400_parquet_file_triggers.sql#L26

vs

https://github.com/influxdata/influxdb_iox/blob/90af3e238b5c8c36d1b80fd974b584bbb95928c4/iox_catalog/migrations/20220913162700_parquetfile_filesize_trigger.sql#L41

### CODE REVIEW

File: iox_catalog/src/sqlite.rs [hunk 1]
```diff
@@ -0,0 +1,2932 @@
...
+    #[tokio::test]
+    async fn test_catalog() {
+        let sqlite = setup_db().await;
+
+        // Validate the connection time zone is the expected UTC value.
+        let tz: Vec<String> = sqlx::query_scalar("pragma module_list;")
+            .fetch_all(&sqlite.pool)
+            .await
+            .expect("read module_list");
+        println!("{}", tz.join(", "));
```
Comment (domodwyer) [inline iox_catalog/src/sqlite.rs:2373]: Should this be asserting the time zone?

File: iox_catalog/src/sqlite.rs [hunk 2]
```diff
@@ -0,0 +1,2932 @@
...
+            .get_observer(&Attributes::from(&[("op", name), ("result", "success")]))
+            .expect("failed to get observer")
+            .fetch();
+
+        let hit_count = histogram.sample_count();
+        assert!(hit_count > 0, "metric did not record any calls");
+    }
+
+    async fn setup_db() -> SqliteCatalog {
+        // TODO(sgc): This is temporary whilst developing SQLite support
```
Comment (domodwyer) [inline iox_catalog/src/sqlite.rs:2352]: I think a separate DNS env & defaulting to an in-memory DB is a good idea going forward too! :+1:

File: clap_blocks/src/catalog_dsn.rs [hunk 3]
```diff
@@ -151,6 +155,21 @@ impl CatalogDsnConfig {
...
+        info!("Catalog: SQLite at `{}`", dsn);
+
+        Self {
+            catalog_type_: CatalogType::Sqlite,
+            dsn: Some(dsn),
+            max_catalog_connections: PostgresConnectionOptions::DEFAULT_MAX_CONNS,
+            postgres_schema_name: PostgresConnectionOptions::DEFAULT_SCHEMA_NAME.to_string(),
+            connect_timeout: PostgresConnectionOptions::DEFAULT_CONNECT_TIMEOUT,
+            idle_timeout: PostgresConnectionOptions::DEFAULT_IDLE_TIMEOUT,
+            hotswap_poll_interval: PostgresConnectionOptions::DEFAULT_HOTSWAP_POLL_INTERVAL,
```
Comment (domodwyer) [inline clap_blocks/src/catalog_dsn.rs:169]: I don't think we'll need the hotswap pool functionality in this implementation - it exists to drive reconnections to postgres using credentials that are rotated by vault - not something that should happen using a local file!

File: iox_catalog/src/sqlite.rs [hunk 4]
```diff
@@ -0,0 +1,2932 @@
...
+        Ok(rec)
+    }
+
+    async fn create_or_get_many_unchecked(
+        &mut self,
+        table_id: TableId,
+        columns: HashMap<&str, ColumnType>,
+    ) -> Result<Vec<Column>> {
+        let num_columns = columns.len();
+        #[derive(Deserialize, Serialize)]
```
Comment (domodwyer) [inline iox_catalog/src/sqlite.rs:809]: Is the JSON encoding here used to inject arrays of values to the query to create multiple rows at once? (as opposed to storing actual JSON in the rows)

File: iox_catalog/Cargo.toml [hunk 5]
```diff
@@ -14,8 +14,10 @@ log = "0.4"
 metric = { version = "0.1.0", path = "../metric" }
 mutable_batch = { path = "../mutable_batch" }
 observability_deps = { path = "../observability_deps" }
+parking_lot = { version = "0.12", features = ["arc_lock"] }
```
Comment (domodwyer) [inline iox_catalog/Cargo.toml:17]: I don't think `arc_lock` is used anywhere?

**Review by domodwyer — APPROVED** [review]
Pretty cool!

**Review by alamb — APPROVED** [review]
I took it for a spin and I love it. Thank you @stuartcarnie !

I only skimmed the catalog code as I believe @domodwyer  is more of an expert in that area and already reviewed. I focused instead on testing this.

What I suggest we do for end to end testing is add a case to all_in_one (I can help doing this if you like) that configures and runs using the sqllite version. Perhaps we can adapt the `smoke` test:

https://github.com/influxdata/influxdb_iox/blob/52b43c40bccd1f47a7e58733d2d6ec5ea51ce081/influxdb_iox/tests/end_to_end_cases/all_in_one.rs#L9-L46


As a follow on PR, I also think we should update the docs for the user facing mode here: https://github.com/influxdata/influxdb_iox/blob/main/README.md#ephemeral-mode as well as tweak the end to end mode user experience . I would be happy to do this



Strangely this did not seem to work (used in memory catalog)

```shell
influxdb_iox run all-in-one -v --catalog=sqlite
```

But this worked great:
```shell
influxdb_iox all-in-one -v --catalog=sqlite --catalog-dsn=/tmp/data.sqlite
```

And I could even check out the database that was created:
```shell
alamb@MacBook-Pro-8:~/Software/influxdb_iox$ sqlite3 /tmp/data.sqlite
SQLite version 3.39.5 2022-10-14 20:58:05
Enter ".help" for usage hints.
sqlite> .tables
_sqlx_migrations        partition               skipped_compactions
billing_summary         processed_tombstone     table_name
column_name             query_pool              tombstone
namespace               shard                   topic
parquet_file            sharding_rule_override
sqlite> select * from column_name limit 10;
1|1|time|6
2|1|host|7
3|1|load1|3
4|1|load15|3
5|1|load5|3
6|1|n_cpus|1
7|1|n_unique_users|1
8|1|n_users|1
9|1|uptime|1
10|1|uptime_format|5
sqlite>
```

**Review by crepererum — APPROVED** [review]
I once hoped that we don't need to rewrite all the SQL queries for this but I think this is unrealistic. SQL as a standard is a lie after all :shrug: 

Anyways, thank you for doing this. I wanted to have a local, file-based catalog for sooo long.

**Comment by stuartcarnie** [comment]
@alamb 

> What I suggest we do for end to end testing is add a case to all_in_one (I can help doing this if you like) that configures and runs using the sqllite version. Perhaps we can adapt the `smoke` test:
> 
> https://github.com/influxdata/influxdb_iox/blob/52b43c40bccd1f47a7e58733d2d6ec5ea51ce081/influxdb_iox/tests/end_to_end_cases/all_in_one.rs#L9-L46
> 

Sounds great!

> As a follow on PR, I also think we should update the docs for the user facing mode here: https://github.com/influxdata/influxdb_iox/blob/main/README.md#ephemeral-mode as well as tweak the end to end mode user experience . I would be happy to do this
> 
> Strangely this did not seem to work (used in memory catalog)
> 
> ```shell
> influxdb_iox run all-in-one -v --catalog=sqlite
> ```
> 

I did a little digging yesterday to determine if we could use an in-memory database for end-to-end testing, however, I found we would have to implement changes, as we spawn the `influxdb_iox` binary to create the catalog:

https://github.com/influxdata/influxdb_iox/blob/dbd27f648fbedfdc5dcc25f87b8a992b76cf0b00/test_helpers_end_to_end/src/database.rs#L30-L38

I wonder if it is a similar issue for `all-in-one` mode?

> But this worked great:
> 
> ```shell
> influxdb_iox all-in-one -v --catalog=sqlite --catalog-dsn=/tmp/data.sqlite
> ```
> 
> And I could even check out the database that was created:
> 
> ```shell
> alamb@MacBook-Pro-8:~/Software/influxdb_iox$ sqlite3 /tmp/data.sqlite
> SQLite version 3.39.5 2022-10-14 20:58:05
> Enter ".help" for usage hints.
> sqlite> .tables
> _sqlx_migrations        partition               skipped_compactions
> billing_summary         processed_tombstone     table_name
> column_name             query_pool              tombstone
> namespace               shard                   topic
> parquet_file            sharding_rule_override
> sqlite> select * from column_name limit 10;
> 1|1|time|6
> 2|1|host|7
> 3|1|load1|3
> 4|1|load15|3
> 5|1|load5|3
> 6|1|n_cpus|1
> 7|1|n_unique_users|1
> 8|1|n_users|1
> 9|1|uptime|1
> 10|1|uptime_format|5
> sqlite>
> ```

This is exciting – thanks for testing that out.

**Comment by stuartcarnie** [comment]
> I once hoped that we don't need to rewrite all the SQL queries for this but I think this is unrealistic. SQL as a standard is a lie after all 🤷

I think there is an opportunity to share a reasonable amount of code between the implementations. I think that exploration is better in a separate PR, as we are in production now.

> 
> Anyways, thank you for doing this. I wanted to have a local, file-based catalog for sooo long.

👍🏻  Me too!

File: clap_blocks/src/catalog_dsn.rs (hunk 3 above)
Comment (stuartcarnie) [reply clap_blocks/src/catalog_dsn.rs:169]: That is what I saw from the PostgreSQL implementation, but wasn't sure what to do with these values, so I just copied this:

https://github.com/influxdata/influxdb_iox/blob/9d07ff69bfaa81fe8050d0e49e773b50cf900d26/clap_blocks/src/catalog_dsn.rs#L132-L140

The most appropriate is probably just set them all to `default` values or change them to `Option` and let the driver choose the default values if they are not set?

File: iox_catalog/src/sqlite.rs (hunk 4 above)
Comment (stuartcarnie) [reply iox_catalog/src/sqlite.rs:809]: Correct – this allows me to generate an array of JSON objects which provide the column name and column type as input to the query. Unfortunately, SQLite doesn't support the `UNNEST` or `ANY` constructs, but the JSON and `json_each` table-valued function provides a reasonable substitute.

File: iox_catalog/Cargo.toml (hunk 5 above)
Comment (stuartcarnie) [reply iox_catalog/Cargo.toml:17]: 👍🏻  I wasn't sure if we had to replicate the features used in the `workspace-hack` project. I'll remove that – thanks for the heads up!

File: iox_catalog/src/sqlite.rs (hunk 1 above)
Comment (stuartcarnie) [reply iox_catalog/src/sqlite.rs:2373]: Oops, that was just some hacky code to verify I had a connection to SQLite in the early stages of development. SQLite doesn't have a default timezone, so I can remove this now.

File: clap_blocks/src/catalog_dsn.rs (hunk 3 above)
Comment (stuartcarnie) [reply clap_blocks/src/catalog_dsn.rs:169]: I added a `Default` implementation via `derive` and updated both the `Memory` and `Sqlite` catalogs to use `..Self::default()` to initialise the unused values.

**Comment by stuartcarnie** [comment]
Update, found out why `Memory` is the default, as when `catalog-dsn` is `None`:

https://github.com/influxdata/influxdb_iox/blob/52b43c40bccd1f47a7e58733d2d6ec5ea51ce081/influxdb_iox/src/commands/run/all_in_one.rs#L365-L369

**Comment by alamb** [comment]
> For some reason, if --catalog-dsn is omitted, it defaults to the Memory catalog type

Once we have the sqllite version of the catalog, I think the defaults for all-in-one should be evaluated / updated -- aka https://github.com/influxdata/influxdb_iox/pull/6851#pullrequestreview-1285041877

**Comment by stuartcarnie** [comment]
> > For some reason, if --catalog-dsn is omitted, it defaults to the Memory catalog type
> 
> Once we have the sqllite version of the catalog, I think the defaults for all-in-one should be evaluated / updated -- aka [#6851 (review)](https://github.com/influxdata/influxdb_iox/pull/6851#pullrequestreview-1285041877). Perhaps this is something we can do as a follow on PR, no need to do it in this PR

Sounds good – I won't make any other changes besides a few small feedback items in this PR and follow on with an update for improved all-in-one operation

File: iox_catalog/src/sqlite.rs (hunk 2 above)
Comment (stuartcarnie) [reply iox_catalog/src/sqlite.rs:2352]: Update, all-in-one does work for in memory, it was just this logic that defaults to `Memory` if the `catalog-dsn` switch is absent:

https://github.com/influxdata/influxdb_iox/blob/52b43c40bccd1f47a7e58733d2d6ec5ea51ce081/influxdb_iox/src/commands/run/all_in_one.rs#L365-L369

I'll do a follow-on PR to default to `:memory:` when `--catalog=sqlite` is specified

**Comment by alamb** [comment]
Thanks @stuartcarnie  -- I wrote up a discussion for all in one mode defaults here: https://github.com/influxdata/influxdb_iox/issues/6885
